clash-lib/src/Clash/Primitives/Xilinx/ClockGen.hs

```diff
@@ -0,0 +1,126 @@
+{-|
+  Copyright   :  (C) 2023, QBayLogic B.V.
+                     2023, Google Inc.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Blackbox template functions for
+  Clash.Xilinx.ClockGen.{clockWizard,clockWizardDifferential}
+-}
+
+{-# LANGUAGE QuasiQuotes #-}
+
+module Clash.Primitives.Xilinx.ClockGen where
+
+import Clash.Backend                      (Backend)
+import Clash.Netlist.BlackBox.Util        (exprToString)
+import Clash.Netlist.Types
+import Clash.Netlist.Util                 (stripVoid)
+
+import Control.Monad.State                (State)
+import qualified Data.String.Interpolate  as I
+import Data.Text.Prettyprint.Doc.Extra    (Doc)
+
+
+clockWizardTclTF :: TemplateFunction
+clockWizardTclTF =
+  TemplateFunction used valid clockWizardTclTemplate
+ where
+  used = [0..4]
+  valid bbCtx
+    | [_,_,_,(nm,_,_),_,_] <- bbInputs bbCtx
+    , Just _ <- exprToString nm
+    , [(Identifier _ Nothing,Product {})] <- bbResults bbCtx
+    = True
+  valid _ = False
+
+clockWizardTclTemplate
+  :: Backend s
+  => BlackBoxContext
+  -> State s Doc
+clockWizardTclTemplate bbCtx = pure bbText
+ where
+  ((_,stripVoid -> kdIn,_):(_,stripVoid -> kdOut,_):_:(nm,_,_):_) = bbInputs bbCtx
```

Comment:
Unfortunately this style generates warnings on GHC 9.2 (non-exhaustive pattern match). Let's do it like this:

```haskell
clockWizardTemplate bbCtx
  | ((_,stripVoid -> kdIn,_):(_,stripVoid -> kdOut,_):_:(nm,_,_):_) <- bbInputs bbCtx
  , Just compName <- exprToString nm
  , KnownDomain _ clkInPeriod _ _ _ _ <- kdIn
  , KnownDomain _ clkOutPeriod _ _ _ _ <- kdOut
  = ...
  | otherwise
  = error ("clockWizardTclTemplate: bad bbContext: " <> show bbCtx)
```